### PR TITLE
Homework 5

### PR DESCRIPTION
**Pull request by demony**

### CODE REVIEW

File: src/test/java/com/epam/tc/hw5/cucumber/context/TestContext.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
+package com.epam.tc.hw5.cucumber.context;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public final class TestContext {
+
+    private static  TestContext instance;
```
Comment (komyagin) [inline src/test/java/com/epam/tc/hw5/cucumber/context/TestContext.java]: Two spaces

File: src/test/java/com/epam/tc/hw5/PropertiesForTests.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+import java.io.IOException;
+import java.util.Properties;
+
+public class PropertiesForTests {
+
+    private final Properties properties = new Properties();
+
+    public PropertiesForTests() {
+
+        String propertyFilePath = "src//test//resources//configs//wh3-configuration.properties";
```
Comment (komyagin) [inline src/test/java/com/epam/tc/hw5/PropertiesForTests.java]: You should move the address string to the constant

File: src/test/java/com/epam/tc/hw5/cucumber/steps/AssertionStep.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+import java.util.List;
+import java.util.stream.Collectors;
+import org.openqa.selenium.WebElement;
+
+public class AssertionStep extends AbstractStep {
+
+    @Then("{string} page should be opened")
+    public void pageShouldBeOpened(String pageTitle) {
+
+        String actualPageTitle = differentElementsPage.getTitle();
```
Comment (komyagin) [inline src/test/java/com/epam/tc/hw5/cucumber/steps/AssertionStep.java]: Why do you need the empty line here?

File: src/test/java/com/epam/tc/hw5/cucumber/steps/AssertionStep.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    @ParameterType("'(.+)'")
+    public List<String> elementNamesTest(String elementNamesAsString) {
+        return Arrays.stream(elementNamesAsString.split(","))
+                     .map(el -> el.trim())
+                     .collect(Collectors.toList());
+    }
+
+    @Then("Checkboxes {elementNamesTest} should be checked")
+    public void checkboxesShouldBeChecked(List<String> elementNamesTest) {
+
```
Comment (komyagin) [inline src/test/java/com/epam/tc/hw5/cucumber/steps/AssertionStep.java]: Why do you need the blank line here?

File: src/test/java/com/epam/tc/hw5/cucumber/steps/AssertionStep.java [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+    @Then("{int} checkboxes should be displayed on Users Table on User Table Page")
+    public void checkboxes_should_be_displayed_on_users_table_on_user_table_page(Integer expectedAmount) {
+        assertThat(userTablePage.getCheckboxes())
+            .as("Correct amount of checkboxes texts on table page")
+            .hasSize(expectedAmount);
+    }
+
+    @Then("User table should contain following values:")
+    public void userTableShouldContainFollowingValues(DataTable dataTable) {
+
```
Comment (komyagin) [inline src/test/java/com/epam/tc/hw5/cucumber/steps/AssertionStep.java]: Why do you need the blank line here?

File: src/test/java/com/epam/tc/hw5/cucumber/steps/AssertionStep.java [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+        List<String> expectedNames = dataTable.column(1).stream().skip(1).collect(Collectors.toList());
+        List<String> actualUserNames = userTablePage.getElementsAsListOfStrings(userTablePage.getUsernames());
+        assertThat(actualUserNames).as("Check column names in table").isEqualTo(expectedNames);
+
+        List<String> expectedDescription = dataTable.column(2).stream().skip(1).collect(Collectors.toList());
+        List<String> actualDescriptions = userTablePage
+            .getElementsAsListOfStrings(userTablePage.getDescriptionTexts());
+        assertThat(actualDescriptions).as("Check column description in table")
+                                      .isEqualTo(expectedDescription);
+
```
Comment (komyagin) [inline src/test/java/com/epam/tc/hw5/cucumber/steps/AssertionStep.java]: Why do you need the blank line here?

File: src/test/java/com/epam/tc/hw5/pages/HomePage.java [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+import org.openqa.selenium.WebDriver;
+
+public class HomePage extends AbstractBasePage {
+
+    public HomePage(WebDriver webDriver) {
+        super(webDriver);
+    }
+
+    public void open() {
+        open("/jdi-light/index.html");
```
Comment (komyagin) [inline src/test/java/com/epam/tc/hw5/pages/HomePage.java]: I think you should move this string to the constant

File: src/test/java/com/epam/tc/hw5/pages/UserTablePage.java [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+    public List<WebElement> getRowNumbers() {
+        return rowNumbers;
+    }
+
+    public List<String> getElementsAsListOfStrings(List<WebElement> webElementLis) {
+        return webElementLis.stream()
+                            .map(el -> el.getText().replaceAll("\n", " "))
+                            .collect(Collectors.toList());
+
```
Comment (komyagin) [inline src/test/java/com/epam/tc/hw5/pages/UserTablePage.java]: Why do you need the blank line here?

File: src/test/java/com/epam/tc/hw5/pages/UserTablePage.java [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+    public List<WebElement> getLogElements() {
+        return logElements;
+    }
+
+    public String getLogByIndex(Integer logIndex) {
+        return logElements.get(logIndex - 1).getText();
+    }
+
```
Comment (komyagin) [inline src/test/java/com/epam/tc/hw5/pages/UserTablePage.java]: Why do you need the blank line here?

File: src/test/java/com/epam/tc/hw5/pages/components/DifferentElementsComponent.java [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+    private List<WebElement> radioButtons;
+
+    @FindBy(xpath = "//div[@class='colors']/select")
+    private WebElement dropDownColors;
+
+    @FindBy(xpath = "//ul[@class='panel-body-list logs']/li[1]")
+    private WebElement firstLogMessage;
+
+    @FindBy(xpath = "//ul[@class='panel-body-list logs']/li")
+    private List<WebElement> logElements;
```
Comment (komyagin) [inline src/test/java/com/epam/tc/hw5/pages/components/DifferentElementsComponent.java]: There are more suitable locators

File: src/test/java/com/epam/tc/hw5/pages/components/DifferentElementsComponent.java [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+    public WebElement getCheckboxByText(String text) {
+        return findSelectorByName(checkboxes, text);
+    }
+
+    public Boolean clickOnRadioButton(String textInsideRadioButton) {
+        return findAndClickOnElementContainingText(radioButtons, textInsideRadioButton);
+    }
+
+    private boolean findAndClickOnElementContainingText(List<WebElement> webElements, String text) {
+
```
Comment (komyagin) [inline src/test/java/com/epam/tc/hw5/pages/components/DifferentElementsComponent.java]: Why do you need the blank line here?

File: src/test/java/com/epam/tc/hw5/pages/components/DifferentElementsComponent.java [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+    public WebElement getCheckboxByText(String text) {
+        return findSelectorByName(checkboxes, text);
+    }
+
+    public Boolean clickOnRadioButton(String textInsideRadioButton) {
+        return findAndClickOnElementContainingText(radioButtons, textInsideRadioButton);
+    }
+
+    private boolean findAndClickOnElementContainingText(List<WebElement> webElements, String text) {
```
Comment (komyagin) [inline src/test/java/com/epam/tc/hw5/pages/components/DifferentElementsComponent.java]: According to the "Clean code" by R.Martin and the SOLID principle you should divide this method into two methods

File: src/test/java/com/epam/tc/hw5/pages/components/DifferentElementsComponent.java [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+    }
+
+    public List<String> getLogElementsTrimDates() {
+        return logElements
+            .stream()
+            .filter(WebElement::isDisplayed)
+            .map(el -> el.getText().substring(9))
+            .collect(Collectors.toList());
+    }
+
```
Comment (komyagin) [inline src/test/java/com/epam/tc/hw5/pages/components/DifferentElementsComponent.java]: Why do you need the blank line here?

File: src/test/java/com/epam/tc/hw5/pages/components/DifferentElementsComponent.java [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+}
+
```
Comment (komyagin) [inline src/test/java/com/epam/tc/hw5/pages/components/DifferentElementsComponent.java]: Why do you need the blank line here?

File: src/test/java/com/epam/tc/hw5/pages/components/LoginComponent.java [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+    @FindBy(css = "button#login-button")
+    private WebElement loginButton;
+
+    public LoginComponent(WebDriver webDriver) {
+        super(webDriver);
+    }
+
+    public void clickToUserIcon() {
+        userIcon.click();
+        // wait.until(visibilityOf(userIcon)).click();
```
Comment (komyagin) [inline src/test/java/com/epam/tc/hw5/pages/components/LoginComponent.java]: Why do you need commented line here?

File: src/test/java/com/epam/tc/hw5/pages/components/LoginComponent.java [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    public void sendKeysToUserPasswordField(String text) {
+        userPasswordInput.sendKeys(text);
+        // wait.until(visibilityOf(userPasswordInput)).sendKeys(text);
```
Comment (komyagin) [inline src/test/java/com/epam/tc/hw5/pages/components/LoginComponent.java]: And here

File: src/test/java/com/epam/tc/hw5/pages/components/LoginComponent.java [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    public void sendKeysToUserLoginField(String text) {
+        userLoginInput.sendKeys(text);
+        // wait.until(visibilityOf(userLoginInput)).sendKeys(text);
```
Comment (komyagin) [inline src/test/java/com/epam/tc/hw5/pages/components/LoginComponent.java]: And here

File: src/test/java/com/epam/tc/hw5/pages/components/MenuComponent.java [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+
+public class MenuComponent extends AbstractComponent {
+
+    @FindBy(css = ".uui-navigation > li > a")
+    private List<WebElement> headerMenu;
+
+    @FindBy(css = ".dropdown-menu > li > a")
+    private List<WebElement> serviceMenu;
+
+    @FindBy(xpath = "//header//a[contains(., 'User Table ')]")
```
Comment (komyagin) [inline src/test/java/com/epam/tc/hw5/pages/components/MenuComponent.java]: There is more suitable locator

File: src/test/java/com/epam/tc/hw5/pages/components/MenuComponent.java [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+
+    public void clickToHeaderMenuElement(String elementName) {
+        findMenuElementByNameAndClickOnIt(headerMenu, elementName);
+    }
+
+    public void clickToMenuElementInServiceDropDown(String elementName) {
+        findMenuElementByNameAndClickOnIt(serviceMenu, elementName);
+    }
+
+    private void findMenuElementByNameAndClickOnIt(List<WebElement> webElementList, String elementName) {
```
Comment (komyagin) [inline src/test/java/com/epam/tc/hw5/pages/components/MenuComponent.java]: SOLID

File: src/test/java/com/epam/tc/hw5/pages/UserTablePage.java [hunk 20]
```diff
@@ -0,0 +1,95 @@
...
+    private List<WebElement> checkboxes;
+
+    @FindBy(css = "td:first-child")
+    private List<WebElement> rowNumbers;
+
+    @FindBy(css = "#user-table tbody > tr")
+    private List<WebElement> tableRows;
+
+    @FindBy(xpath = "//ul[@class='panel-body-list logs']/li")
+    private List<WebElement> logElements;
```
Comment (komyagin) [inline src/test/java/com/epam/tc/hw5/pages/UserTablePage.java]: There are more suitable locators

**Review by komyagin — CHANGES_REQUESTED** [review]

File: src/test/java/com/epam/tc/hw5/cucumber/context/TestContext.java (hunk 1 above)
Comment (demony) [reply src/test/java/com/epam/tc/hw5/cucumber/context/TestContext.java]: Fixed.

File: src/test/java/com/epam/tc/hw5/PropertiesForTests.java (hunk 2 above)
Comment (demony) [reply src/test/java/com/epam/tc/hw5/PropertiesForTests.java]: Fixed.

File: src/test/java/com/epam/tc/hw5/cucumber/steps/AssertionStep.java (hunk 3 above)
Comment (demony) [reply src/test/java/com/epam/tc/hw5/cucumber/steps/AssertionStep.java]: Fixed.

File: src/test/java/com/epam/tc/hw5/cucumber/steps/AssertionStep.java (hunk 4 above)
Comment (demony) [reply src/test/java/com/epam/tc/hw5/cucumber/steps/AssertionStep.java]: Fixed.

File: src/test/java/com/epam/tc/hw5/cucumber/steps/AssertionStep.java (hunk 5 above)
Comment (demony) [reply src/test/java/com/epam/tc/hw5/cucumber/steps/AssertionStep.java]: Fixed.

File: src/test/java/com/epam/tc/hw5/pages/components/LoginComponent.java (hunk 16 above)
Comment (demony) [reply src/test/java/com/epam/tc/hw5/pages/components/LoginComponent.java]: Fixed.

File: src/test/java/com/epam/tc/hw5/cucumber/steps/AssertionStep.java (hunk 6 above)
Comment (demony) [reply src/test/java/com/epam/tc/hw5/cucumber/steps/AssertionStep.java]: Fixed.

File: src/test/java/com/epam/tc/hw5/pages/HomePage.java (hunk 7 above)
Comment (demony) [reply src/test/java/com/epam/tc/hw5/pages/HomePage.java]: Fixed.

File: src/test/java/com/epam/tc/hw5/pages/UserTablePage.java (hunk 8 above)
Comment (demony) [reply src/test/java/com/epam/tc/hw5/pages/UserTablePage.java]: Fixed.

File: src/test/java/com/epam/tc/hw5/pages/UserTablePage.java (hunk 9 above)
Comment (demony) [reply src/test/java/com/epam/tc/hw5/pages/UserTablePage.java]: Fixed.

File: src/test/java/com/epam/tc/hw5/pages/components/DifferentElementsComponent.java (hunk 10 above)
Comment (demony) [reply src/test/java/com/epam/tc/hw5/pages/components/DifferentElementsComponent.java]: Fixed.

File: src/test/java/com/epam/tc/hw5/pages/components/DifferentElementsComponent.java (hunk 11 above)
Comment (demony) [reply src/test/java/com/epam/tc/hw5/pages/components/DifferentElementsComponent.java]: Fixed.

File: src/test/java/com/epam/tc/hw5/pages/components/DifferentElementsComponent.java (hunk 13 above)
Comment (demony) [reply src/test/java/com/epam/tc/hw5/pages/components/DifferentElementsComponent.java]: Fixed.

File: src/test/java/com/epam/tc/hw5/pages/components/DifferentElementsComponent.java (hunk 14 above)
Comment (demony) [reply src/test/java/com/epam/tc/hw5/pages/components/DifferentElementsComponent.java]: Fixed.

File: src/test/java/com/epam/tc/hw5/pages/components/LoginComponent.java (hunk 15 above)
Comment (demony) [reply src/test/java/com/epam/tc/hw5/pages/components/LoginComponent.java]: Fixed.

File: src/test/java/com/epam/tc/hw5/pages/components/LoginComponent.java (hunk 17 above)
Comment (demony) [reply src/test/java/com/epam/tc/hw5/pages/components/LoginComponent.java]: Fixed.

File: src/test/java/com/epam/tc/hw5/pages/components/MenuComponent.java (hunk 18 above)
Comment (demony) [reply src/test/java/com/epam/tc/hw5/pages/components/MenuComponent.java]: Fixed.

File: src/test/java/com/epam/tc/hw5/pages/components/MenuComponent.java (hunk 19 above)
Comment (demony) [reply src/test/java/com/epam/tc/hw5/pages/components/MenuComponent.java]: Fixed.

File: src/test/java/com/epam/tc/hw5/pages/UserTablePage.java (hunk 20 above)
Comment (demony) [reply src/test/java/com/epam/tc/hw5/pages/UserTablePage.java]: Fixed. Except 
    @FindBy(css = "td:first-child")
    private List<WebElement> rowNumbers;
Can't find solution :(

**Review by komyagin — APPROVED** [review]

**Pull request merged**